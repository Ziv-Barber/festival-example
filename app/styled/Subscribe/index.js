// @flow

import React from 'react' // eslint-disable-line no-unused-vars
import styled from 'styled-components'

// Other components that we need:
import TextInput from '../TextInput'

// Material-ui stuff:
import MuiButton from '@material-ui/core/Button'
import MuiCheckbox from '@material-ui/core/Checkbox'
import MuiFormControlLabel from '@material-ui/core/FormControlLabel'
// import MuiPaper from '@material-ui/core/Paper'
// import MuiTypography from '@material-ui/core/Typography'
// import MuiGrid from '@material-ui/core/Grid'

// Our types:
import type { SubscribePropsType } from './index.types'

/**
 * Subscribe react component.
 * @reactProps {string} text Parameters for the component.
 * @return New element.
 */
const Subscribe = (props: SubscribePropsType) => {
  const [email, setEmail] = React.useState('')
  const [emailErr, setEmailErr] = React.useState(false)
  const [firstName, setFirstName] = React.useState('')
  const [firstNameErr, setFirstNameErr] = React.useState(false)
  const [lastName, setLastName] = React.useState('')
  const [lastNameErr, setLastNameErr] = React.useState(false)
  const [accept, setAccept] = React.useState('')
  const [acceptErr, setAcceptErr] = React.useState(false)

  const handleChange = (
    setValue: (string | boolean) => void,
    setErr: (boolean) => void,
    isValidateEmail: boolean = false,
    isCheckBox: boolean = false
  ) => (event: { target: { value: string, checked?: boolean } }) => {
    const recvValue = isCheckBox ? event.target.checked : event.target.value

    if (typeof recvValue === 'string' && recvValue.length < 1) {
      if (typeof recvValue === 'string' || typeof recvValue === 'boolean') {
        setValue(recvValue)
      } // Endif.

      setErr(true)
      return
    } else if (recvValue === false) {
      if (typeof recvValue === 'string' || typeof recvValue === 'boolean') {
        setValue(recvValue)
      } // Endif.

      setErr(true)
      return
    } // Endif.

    // In case that this is an email field - check if the email is valid:
    const emailValidator = /^([^\x00-\x20\x22\x28\x29\x2c\x2e\x3a-\x3c\x3e\x40\x5b-\x5d\x7f-\xff]+|\x22([^\x0d\x22\x5c\x80-\xff]|\x5c[\x00-\x7f])*\x22)(\x2e([^\x00-\x20\x22\x28\x29\x2c\x2e\x3a-\x3c\x3e\x40\x5b-\x5d\x7f-\xff]+|\x22([^\x0d\x22\x5c\x80-\xff]|\x5c[\x00-\x7f])*\x22))*\x40([^\x00-\x20\x22\x28\x29\x2c\x2e\x3a-\x3c\x3e\x40\x5b-\x5d\x7f-\xff]+|\x5b([^\x0d\x5b-\x5d\x80-\xff]|\x5c[\x00-\x7f])*\x5d)(\x2e([^\x00-\x20\x22\x28\x29\x2c\x2e\x3a-\x3c\x3e\x40\x5b-\x5d\x7f-\xff]+|\x5b([^\x0d\x5b-\x5d\x80-\xff]|\x5c[\x00-\x7f])*\x5d))*$/ // eslint-disable-line no-control-regex
    if (isValidateEmail && !emailValidator.test(String(email).toLowerCase())) {
      if (typeof recvValue === 'string' || typeof recvValue === 'boolean') {
        setValue(recvValue)
      } // Endif.

      setErr(true)
      return
    } // Endif.

    if (typeof recvValue === 'string' || typeof recvValue === 'boolean') {
      setValue(recvValue)
    } // Endif.

    setErr(false)
  }

  return (
    <div className={props.className}>
      <TextInput
        id="email"
        className="textInput"
        label="Email"
        value={email}
        error={emailErr || undefined}
        onChange={handleChange(setEmail, setEmailErr, true)}
      />
      <TextInput
        id="first"
        className="textInput"
        label="First Name"
        value={firstName}
        error={firstNameErr || undefined}
        onChange={handleChange(setFirstName, setFirstNameErr)}
      />
      <TextInput
        id="last"
        className="textInput"
        label="Last Name"
        value={lastName}
        error={lastNameErr || undefined}
        onChange={handleChange(setLastName, setLastNameErr)}
      />
      <MuiFormControlLabel
        control={
          <MuiCheckbox
            checked={!!accept}
            onChange={handleChange(setAccept, setAcceptErr, false, true)}
            value="accepted"
            error={acceptErr || undefined}
          />
        }
        label="Yes, I would like to receive the free newsletter per email. I can withdraw my consent at any time. I've read the Privacy Policy."
      />
      <MuiButton
        variant="contained"
        color="primary"
        className="subButton"
        disabled={
          !email ||
          emailErr ||
          !firstName ||
          firstNameErr ||
          !lastName ||
          lastNameErr ||
          !accept
        }
      >
        Subscribe
      </MuiButton>
    </div>
  )
}

export default styled(Subscribe)`
  padding: ${(props) => props.theme.spacing.unit}px !important;
  padding-top: 0 !important;
  padding-bottom: ${(props) => props.theme.spacing.unit * 3}px !important;
  background: ${(props) => props.theme.palette.background.default};
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;

  & .textInput {
    width: 100%;
  }

  @media (min-width: 600px) {
    & .textInput {
      width: 580px;
    }
  }

  & .subButton {
  }
`

// In case that someone need our types:
export type { SubscribePropsType as Props }
