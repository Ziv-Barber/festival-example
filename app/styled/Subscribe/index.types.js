// @flow

export type SubscribePropsType = {
  className?: string
}
