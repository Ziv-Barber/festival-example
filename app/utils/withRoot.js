// @flow

import * as React from 'react' // eslint-disable-line no-unused-vars
import { MuiThemeProvider } from '@material-ui/core/styles'
import { ThemeProvider } from 'styled-components'
import CssBaseline from '@material-ui/core/CssBaseline'
import theme from '../theme'

/**
 * withRoot HOC to wrap a react component with the app's theme provider.
 */
const withRoot = (Component: React.ComponentType<any>) => {
  const withRootBase = (props: any) => (
    <MuiThemeProvider theme={theme}>
      <ThemeProvider theme={theme}>
        <React.Fragment>
          <CssBaseline />
          <Component {...props} />
        </React.Fragment>
      </ThemeProvider>
    </MuiThemeProvider>
  )

  // $FlowFixMe
  withRootBase.displayName = `withRoot(${Component.displayName ||
    Component.name})`

  return withRootBase
}

export default withRoot
