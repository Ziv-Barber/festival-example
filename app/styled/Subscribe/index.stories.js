import React from 'react' // eslint-disable-line no-unused-vars
import { storiesOf } from '@storybook/react'
import Subscribe from './index'

storiesOf('Subscribe', module).add(
  'basic',
  () => <Subscribe text="Basic example" />,
  { notes: { markdown: 'This is an example to the Subscribe component.' } }
)
