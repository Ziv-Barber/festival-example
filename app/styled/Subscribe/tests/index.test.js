// @flow

import React from 'react' // eslint-disable-line no-unused-vars
import { render, cleanup } from 'react-testing-library'

import Subscribe from '../index'
import withRoot from '../../../utils/withRoot'

// For flow:
let { describe, it, afterEach, expect } = global

/** @test {Subscribe} */
describe('<Subscribe />', () => {
  afterEach(cleanup)

  /** @test {Subscribe} */
  it('should match snapshort', () => {
    const App = withRoot(Subscribe)
    const { container } = render(<App />)
    expect(container).toMatchSnapshot()
  })
})
