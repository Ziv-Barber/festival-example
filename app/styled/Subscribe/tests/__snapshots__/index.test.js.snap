// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Subscribe /> should match snapshort 1`] = `
<div>
  <div
    class="Subscribe-sc-4khb8h-0 hCaqfs"
  >
    <div
      class="MuiFormControl-root-1 MuiFormControl-marginNormal-2 textInput TextInput-k6q9l7-0 jjzSuf"
    >
      <label
        class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink="false"
        for="email"
      >
        Email
      </label>
      <div
        class="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
      >
        <fieldset
          aria-hidden="true"
          class="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style="padding-left: 8px;"
        >
          <legend
            class="MuiPrivateNotchedOutline-legend-54"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
        <input
          aria-invalid="false"
          class="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          id="email"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-1 MuiFormControl-marginNormal-2 textInput TextInput-k6q9l7-0 jjzSuf"
    >
      <label
        class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink="false"
        for="first"
      >
        First Name
      </label>
      <div
        class="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
      >
        <fieldset
          aria-hidden="true"
          class="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style="padding-left: 8px;"
        >
          <legend
            class="MuiPrivateNotchedOutline-legend-54"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
        <input
          aria-invalid="false"
          class="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          id="first"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-1 MuiFormControl-marginNormal-2 textInput TextInput-k6q9l7-0 jjzSuf"
    >
      <label
        class="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink="false"
        for="last"
      >
        Last Name
      </label>
      <div
        class="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
      >
        <fieldset
          aria-hidden="true"
          class="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style="padding-left: 8px;"
        >
          <legend
            class="MuiPrivateNotchedOutline-legend-54"
            style="width: 0.01px;"
          >
            <span>
              ​
            </span>
          </legend>
        </fieldset>
        <input
          aria-invalid="false"
          class="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          id="last"
          type="text"
          value=""
        />
      </div>
    </div>
    <label
      class="MuiFormControlLabel-root-55"
    >
      <span
        class="MuiButtonBase-root-77 MuiIconButton-root-71 MuiPrivateSwitchBase-root-67 MuiCheckbox-root-61 MuiCheckbox-colorSecondary-66"
      >
        <span
          class="MuiIconButton-label-76"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-80"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <input
            class="MuiPrivateSwitchBase-input-70"
            data-indeterminate="false"
            type="checkbox"
            value="accepted"
          />
        </span>
        <span
          class="MuiTouchRipple-root-151"
        />
      </span>
      <span
        class="MuiTypography-root-89 MuiTypography-body2-97 MuiFormControlLabel-label-60"
      >
        Yes, I would like to receive the free newsletter per email. I can withdraw my consent at any time. I've read the Privacy Policy.
      </span>
    </label>
    <button
      class="MuiButtonBase-root-77 MuiButtonBase-disabled-78 MuiButton-root-125 MuiButton-contained-136 MuiButton-containedPrimary-137 MuiButton-raised-139 MuiButton-raisedPrimary-140 MuiButton-disabled-145 subButton"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label-126"
      >
        Subscribe
      </span>
    </button>
  </div>
</div>
`;
